core/lib/dal/src/consensus_dal.rs

```diff
@@ -30,54 +30,129 @@ impl ConsensusDal<'_, '_> {
+                got.fork.number,
+                genesis.fork.number,
+            );
+            anyhow::ensure!(
+                got.fork.first_parent.is_none(),
+                "fork with first_parent != None not supported",
+            );
+        }
+        let genesis =
+            zksync_protobuf::serde::serialize(genesis, serde_json::value::Serializer).unwrap();
+        let state = zksync_protobuf::serde::serialize(
+            &ReplicaState::default(),
+            serde_json::value::Serializer,
+        )
+        .unwrap();
+        sqlx::query!(
+            r#"
+            DELETE FROM miniblocks_consensus
+            "#
+        )
+        .execute(txn.conn())
+        .await?;
+        sqlx::query!(
+            r#"
+            DELETE FROM consensus_replica_state
+            "#
+        )
+        .execute(txn.conn())
+        .await?;
         sqlx::query!(
             r#"
             INSERT INTO
-                consensus_replica_state (fake_key, state)
+                consensus_replica_state (fake_key, genesis, state)
             VALUES
-                (TRUE, $1)
-            ON CONFLICT (fake_key) DO
-            UPDATE
-            SET
-                state = excluded.state
+                (TRUE, $1, $2)
             "#,
-            state
+            genesis,
+            state,
         )
-        .execute(self.storage.conn())
+        .execute(txn.conn())
         .await?;
+        txn.commit().await?;
         Ok(())
     }
 
-    /// Fetches the first consensus certificate.
-    /// Note that we didn't backfill the certificates for the past miniblocks
-    /// when enabling consensus certificate generation, so it might NOT be the certificate
-    /// for the genesis miniblock.
-    pub async fn first_certificate(&mut self) -> anyhow::Result<Option<validator::CommitQC>> {
-        let Some(row) = sqlx::query!(
+    /// [Main node only] creates a new consensus fork starting at
+    /// the last sealed miniblock. Resets the state of the consensus
+    /// by calling `try_update_genesis()`.
+    pub async fn fork(&mut self) -> anyhow::Result<()> {
+        let mut txn = self.storage.start_transaction().await?;
+        let Some(old) = txn.consensus_dal().genesis().await? else {
+            return Ok(());
+        };
+        let last = txn.blocks_dal().get_sealed_miniblock_number().await?;
+        let first_block = validator::BlockNumber(last.map(|n| n.0.into()).unwrap_or(0));
+
+        let new = validator::Genesis {
+            validators: old.validators,
+            fork: validator::Fork {
+                number: old.fork.number.next(),
+                first_block,
+                first_parent: None,
+            },
+        };
+        txn.consensus_dal().try_update_genesis(&new).await?;
+        txn.commit().await?;
+        Ok(())
+    }
+
+    /// Fetches the current BFT replica state.
+    pub async fn replica_state(&mut self) -> anyhow::Result<ReplicaState> {
```

Comment:
Dumb question: Non-optionality of the returned state is caused by the changes in the consensus logic (e.g., replica state never being read before it's written), right? From the DAL perspective, I don't think it's reasonable to assume that the `consensus_replica_state` table is not empty. Sure, we won't panic if it is, but IMO, it'd be reasonable to return `Result<Option<_>>` and deal with `Ok(None)` on the app level.